snapcraft/plugins/dump.py

```diff
@@ -35,7 +35,58 @@ class DumpPlugin(snapcraft.BasePlugin):
 
     def build(self):
         super().build()
-        if os.path.exists(self.installdir):
-            os.rmdir(self.installdir)
-        shutil.copytree(self.builddir, self.installdir,
-                        copy_function=snapcraft.common.link_or_copy)
+        _linktree(self.builddir, self.installdir, self.installdir)
+
+
+def _link_or_copy(source, destination, boundary):
+    """Attempt to copy symlinks as symlinks unless pointing out of boundary."""
+
+    follow_symlinks = False
+
+    # If this is a symlink, analyze where it's pointing and make sure it will
+    # still be valid when snapped. If it won't, follow the symlink when
+    # copying (i.e. copy the file to which the symlink is pointing instead).
+    if os.path.islink(source):
+        link = os.readlink(source)
+        destination_dirname = os.path.dirname(destination)
+        normalized = os.path.normpath(os.path.join(destination_dirname, link))
+        if os.path.isabs(link) or not normalized.startswith(boundary):
+            follow_symlinks = True
+
+    try:
+        snapcraft.common.link_or_copy(source, destination,
+                                      follow_symlinks=follow_symlinks)
+    except FileNotFoundError:
+        raise FileNotFoundError('{!r} is a broken symlink pointing outside the snap'.format(source))
+
+
+def _create_similar_directory(source, destination):
+    os.makedirs(destination, exist_ok=True)
+    shutil.copystat(source, destination, follow_symlinks=False)
+
+
+def _linktree(source_tree, destination_tree, boundary):
```

Comment:
on the first read, the meaning of the boundary argument is not clear. A comment would be nice.